File: src/main/java/workflow/process/services/FileProcessService.java
package workflow.process.services;


import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import workflow.process.data.model.*;
import workflow.process.exception.DataPersistenceException;

import java.util.List;
import java.util.UUID;

@Service
@Slf4j
@RequiredArgsConstructor
public class FileProcessService {

    private final DataPersistenceService dataPersistenceService;
    private final FileProcessStorageService fileProcessStorageService;

    @NotNull
    public FileProcessDto processFile(@NotNull final MultipartFile fileToProcess) {
        String s3Path = null;
        try {
            validateFile(fileToProcess);

            s3Path = fileProcessStorageService.pushToS3(fileToProcess);

            final FileProcess newFileProcess = constructFileProcess(s3Path, fileToProcess);
            final Outbox outbox = constructOutbox(newFileProcess);

            final FileProcess upsertedFileProcess = dataPersistenceService.persistData(newFileProcess, outbox);
            return new FileProcessDto(upsertedFileProcess.getUuid(), upsertedFileProcess.getStatus().toString(), upsertedFileProcess.getUpdatedAt().toString());
        } catch (DataPersistenceException dataPersistenceException) {
            log.error("Failed to persist data: ", dataPersistenceException);
            cleanUpS3(s3Path);
            throw new RuntimeException("Failed to process file", dataPersistenceException);
        } catch (Exception e) {
            throw new RuntimeException("Failed to process file", e);
        }
    }

    @NotNull
    public List<FileProcessDto> fetchAllFileProcesses() {
        return dataPersistenceService.fetchAllFileProcesses();
    }

    private void cleanUpS3(@Nullable final String s3Path) {
        if (s3Path != null) {
            fileProcessStorageService.deleteFromS3(s3Path);
        }
    }

    private Outbox constructOutbox(@NotNull final FileProcess fileProcess) {
        final Outbox outbox = new Outbox();
        outbox.setFileUUID(fileProcess.getUuid());
        outbox.setStatus(OutboxStatus.PENDING);
        return outbox;
    }

    private FileProcess constructFileProcess(@NotNull final String s3Path,
                                             @NotNull final MultipartFile fileToProcess) {
        final FileProcess fileProcess = new FileProcess();
        fileProcess.setFileName(fileToProcess.getOriginalFilename());
        fileProcess.setUuid(UUID.randomUUID().toString());
        fileProcess.setStatus(FileStatus.PROCESSING);
        fileProcess.setFileS3Path(s3Path);
        return fileProcess;
    }


    /**
     * Perform basic file validations
     */
    private static void validateFile(MultipartFile fileToProcess) {
        if (fileToProcess.getOriginalFilename() == null) {
            throw new RuntimeException("File name is required");
        }
        if (fileToProcess.getSize() >= 9000000) {
            throw new RuntimeException("File size is too large");
        }
    }

}

File: src/main/kotlin/workflow/process/services/FileProcessService.kt
package workflow.process.services


import jakarta.persistence.OptimisticLockException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import workflow.process.data.FileStatus
import workflow.process.data.model.FileProcess
import workflow.process.data.model.FileProcessDto
import workflow.process.events.FileEventProducer
import workflow.process.extensions.toDto
import workflow.process.extensions.toEntity

@Service
class FileProcessService(
    private val fileProcessRepository: workflow.process.data.FileProcessRepository,
    private val fileEventProducer: FileEventProducer,
    private val fileProcessStorageService: FileProcessStorageService
) {

    @Transactional(rollbackFor = [Exception::class])
    fun upsertFileProcess(multipartFile: MultipartFile): FileProcessDto {
        val fileName = multipartFile.originalFilename ?: throw IllegalArgumentException("File name is required")
        checkFileNameNotInUse(fileName)

        // Step1: Create initial entry in DB
        val fileProcess = save(
            createFileProcess(fileName, null, FileStatus.QUEUED).toEntity()
        ).toDto()

        try {

            // Step2: Upload on s3
            val s3Url = fileProcessStorageService.pushToS3(multipartFile);
            fileProcess.status = FileStatus.UPLOADED
            fileProcess.s3Url = s3Url

            // Step3: After s3 upload, we upsert the entry back with s3 url and UPLOADED status
            save(fileProcess.toEntity())


            // Step4: We enqueue event to amq. Status PROCESSING will be done by CONSUMER
            enqueueEventInAmqp(fileProcess.fileName)

        } catch (e: Exception) {
            throw e;
        }
        return fileProcess
    }


    private fun createFileProcess(fileName: String, s3Url: String?, status: FileStatus): FileProcessDto {
        val fileProcess = FileProcessDto(
            fileName = fileName,
            s3Url = s3Url,
            status = status
        )
        return fileProcess
    }

    private fun checkFileNameNotInUse(fileName: String) {
        val existingFileProcess = findByName(fileName)
        if (existingFileProcess != null) {
            throw IllegalArgumentException("File name is already in use.")
        }
    }

    fun enqueueEventInAmqp(fileName: String) {
        fileEventProducer.produceFileEvent(fileName)
    }

    fun save(fileProcess: FileProcess): FileProcess {
        try {
            val existingFileProcess = findByName(fileProcess.fileName!!)
            if (existingFileProcess != null) {
                fileProcess.id = existingFileProcess.id
                return fileProcessRepository.save(fileProcess)
            }
            return fileProcessRepository.save(fileProcess)
        } catch (e: OptimisticLockException) {
            throw RuntimeException("Error uploading file process.")
        }
    }

    fun findByName(fileName: String): FileProcess? {
        return fileProcessRepository.findByFileName(fileName)
    }

    fun findById(id: Int): FileProcess {
        return fileProcessRepository.findById(id).orElseThrow { RuntimeException("FileProcess not found") }
    }

    fun findAll(): List<FileProcess> {
        return fileProcessRepository.findAll().toCollection(ArrayList())
    }

}
